File: programmers/algo/070000+/PG_72410_신규_아이디_추천.cpp
// Solve 2023-03-21
// Update 2025-04-16

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

string solution(string new_id) {
    string answer = new_id;

    // Step 1

    transform(answer.begin(), answer.end(), answer.begin(), ::tolower);

    // Step 2

    for (int i = 0; i < answer.length(); i++) {
        if (islower(answer[i]) || isdigit(answer[i]) || answer[i] == '-' || answer[i] == '_' || answer[i] == '.') {
            continue;
        }

        answer.erase(answer.begin() + i);
        i--;
    }

    // Step 3

    for (int i = 1; i < answer.length(); i++) {
        if (answer[i] == '.' && answer[i - 1] == '.') {
            answer.erase(answer.begin() + i);
            i--;
        }
    }

    // Step 4

    while (!answer.empty() && answer[0] == '.') {
        answer.erase(answer.begin());
    }

    while (!answer.empty() && answer.back() == '.') {
        answer.pop_back();
    }

    // Step 5

    if (answer.empty()) {
        answer = "a";
    }

    // Step 6

    if (answer.size() >= 16) {
        answer.erase(answer.begin() + 15, answer.end());
    }

    while (!answer.empty() && answer.back() == '.') {
        answer.pop_back();
    }

    // Step 7

    while (answer.size() <= 2) {
        answer.push_back(answer.back());
    }

    return answer;
}

File: programmers/algo/070000+/PG_72410_신규_아이디_추천.java
// Solve 2025-04-16

class PG_72410_신규_아이디_추천 {

    boolean isLower(char ch) {
        return 'a' <= ch && ch <= 'z';
    }

    boolean isDigit(char ch) {
        return '0' <= ch && ch <= '9';
    }

    boolean isValidCharacter(char ch) {
        return isLower(ch) || isDigit(ch) || ch == '-' || ch == '_' || ch == '.';
    }

    String removeContinuousDot(String str) {
        StringBuilder sb = new StringBuilder();

        for (char ch : str.toCharArray()) {
            if (ch == '.' && sb.length() >= 1 && sb.charAt(sb.length() - 1) == '.') continue;

            sb.append(ch);
        }

        return sb.toString();
    }

    String removeBeginningDot(String str) {
        StringBuilder sb = new StringBuilder();
        int i_begin = 0;

        while (i_begin < str.length() && str.charAt(i_begin) == '.') {
            i_begin++;
        }

        for (int i = i_begin; i < str.length(); i++) {
            sb.append(str.charAt(i));
        }

        return sb.toString();
    }

    String removeEndingDot(String str) {
        StringBuilder sb = new StringBuilder();
        int i_end = str.length() - 1;

        while (i_end >= 0 && str.charAt(i_end) == '.') {
            i_end--;
        }

        for (int i = 0; i <= i_end; i++) {
            sb.append(str.charAt(i));
        }

        return sb.toString();
    }

    public String solution(String new_id) {
        // Step 1

        new_id = new_id.toLowerCase();

        // Step 2

        StringBuilder sb = new StringBuilder();

        for (char ch : new_id.toCharArray()) {
            if (isValidCharacter(ch)) {
                sb.append(ch);
            }
        }

        new_id = sb.toString();

        // Step 3

        new_id = removeContinuousDot(new_id);

        // Step 4

        new_id = removeBeginningDot(new_id);
        new_id = removeEndingDot(new_id);

        // Step 5

        if (new_id.isEmpty()) {
            new_id = "a";
        }

        // Step 6

        if (new_id.length() >= 16) {
            new_id = new_id.substring(0, 15);
            new_id = removeEndingDot(new_id);
        }

        // Step 7

        while (new_id.length() <= 2) {
            new_id += new_id.charAt(new_id.length() - 1);
        }

        return new_id;
    }

}
